### PR TITLE
[AMDAIEControlCodeLowering] Separate DmaCpyNdToHalfDmaCpyNd to a new pass

### PR DESCRIPTION
**Pull request by Yu-Zhewen**
- Introduce a new pass that specifically handles the conversion of `DmaCpyNd` to `HalfDmaCpyNd`.
- The remaining logic in `ControlCodeLowering`  focuses on further lowering `HalfDmaCpyNd` into `WriteBd`, `AddressPatch`, and `PushToQueue`.
- These changes are required for enabling DMA BD chaining #931

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/controlcode_lowering.mlir [hunk 1]
```diff
@@ -58,19 +58,22 @@ module attributes {hal.executable.target = #executable_target_amdaie_xclbin_fb}
 // CHECK: amdaie.npu.write_bd {bd_id = 0 : ui32, buffer_length = 0 : ui32, buffer_offset = 0 : ui32, col = 0 : ui32, enable_packet = true, iteration_current = 0 : ui32, iteration_size = 0 : ui32, iteration_stride = 0 : ui32, lock_acq_enable = false, lock_acq_id = 0 : ui32, lock_acq_val = 0 : i32, lock_rel_id = 0 : ui32, lock_rel_val = 0 : i32, next_bd = 0 : ui32, out_of_order_id = 0 : ui32, packet_id = 0 : ui32, packet_type = 0 : ui32, paddings_after = array<i32>, paddings_before = array<i32>, row = 0 : ui32, sizes = array<i32: 0, 0, 0>, strides = array<i32: 0, 0, 0>, use_next_bd = false, valid_bd = true}
 // CHECK: amdaie.npu.address_patch {arg_idx = 0 : ui32, bd_id = 0 : ui32, col = 0 : ui32, offset = 0 : ui32}
 // CHECK: amdaie.npu.push_to_queue {bd_id = 0 : ui32, channel = 0 : ui32, col = 0 : ui32, direction = 1 : i32, repeat_count = 1 : ui32, row = 0 : ui32}
-        amdaie.npu.dma_cpy_nd %4([] [] [], %5[] [] [] bd_id = %bd_id) : source_type = !amdaie.logicalobjectfifo<memref<2048xi32>>
+        amdaie.npu.half_dma_cpy_nd %4(%5 [] [] [] bd_id = %bd_id channel = %channel) : !amdaie.logicalobjectfifo<memref<2048xi32>>
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/controlcode_lowering.mlir:61]: The names of these tests are still `npu_dma_cpy_nd...`, could you rename to `half_npu_dma_cpy_nd...`?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/controlcode_lowering.mlir [hunk 2]
```diff
@@ -211,120 +220,20 @@ module attributes {hal.executable.target = #executable_target_amdaie_xclbin_fb}
...
-        amdaie.end
-      }
-    }
-    return
-  }
-}
-
-// -----
-
-// CHECK-LABEL: @half_npu_dma_cpy_nd
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/controlcode_lowering.mlir:235]: Not sure why you're removing the `half_npu_dma_cpy_nd` tests and keeping the `npu_dma_cpy_nd` ones? I would have expected it the other way around? And unless they are exactly duplicates, you could also keep both of them?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/npu_dma_to_half_dma_cpy_nd.mlir [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+// CHECK: amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
+// CHECK: %[[TOKEN_0:.+]] = amdaie.npu.half_dma_cpy_nd async %[[CONNECTION]](%[[OBJECT_FIFO_1]] [0, 0, 0, 0] [2, 4, 16, 16] [0, 64, 8, 1] bd_id = %[[BD_ID]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+// CHECK: amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
+// CHECK: amdaie.npu.dma_wait(%[[TOKEN_0]] : !amdaie.async_token)
+// CHECK: %[[TOKEN_1:.+]] = amdaie.npu.half_dma_cpy_nd async %[[CONNECTION]](%[[OBJECT_FIFO_1]] [0, 0, 0, 32] [2, 4, 16, 16] [128, 64, 8, 1] bd_id = %[[BD_ID]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+// CHECK: amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
+// CHECK: amdaie.npu.dma_wait(%[[TOKEN_1]] : !amdaie.async_token)
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/npu_dma_to_half_dma_cpy_nd.mlir:46]:
```suggestion
// CHECK-LABEL: @npu_dma_cpy_nd_source
// CHECK:             %[[BD_ID:.+]] = amdaie.bd_id
// CHECK:             %[[OBJECT_FIFO_0:.+]] = amdaie.logicalobjectfifo.from_buffers
// CHECK:             %[[CHANNEL_0:.+]] = amdaie.channel
// CHECK:             %[[CHANNEL_1:.+]] = amdaie.channel
// CHECK:             %[[CONNECTION:.+]] = amdaie.connection
// CHECK:             %[[OBJECT_FIFO_1:.+]] = amdaie.logicalobjectfifo.from_memref
// CHECK:             amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_1]] [] [] [] bd_id = %[[BD_ID]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
// CHECK:             amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
// CHECK:             %[[TOKEN_0:.+]] = amdaie.npu.half_dma_cpy_nd async %[[CONNECTION]](%[[OBJECT_FIFO_1]] [0, 0, 0, 0] [2, 4, 16, 16] [0, 64, 8, 1] bd_id = %[[BD_ID]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
// CHECK:             amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
// CHECK:             amdaie.npu.dma_wait(%[[TOKEN_0]] : !amdaie.async_token)
// CHECK:             %[[TOKEN_1:.+]] = amdaie.npu.half_dma_cpy_nd async %[[CONNECTION]](%[[OBJECT_FIFO_1]] [0, 0, 0, 32] [2, 4, 16, 16] [128, 64, 8, 1] bd_id = %[[BD_ID]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
// CHECK:              amdaie.npu.half_dma_cpy_nd %[[CONNECTION]](%[[OBJECT_FIFO_0]] [] [] [] channel = %[[CHANNEL_1]]) : !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
// CHECK:              amdaie.npu.dma_wait(%[[TOKEN_1]] : !amdaie.async_token)
```
This is a style suggestion, but I think using indentation is more readable. Although we're not exactly consistent on this format, most tests use this indented format.

**Review by jtuyls — COMMENTED** [review]
Thanks! Just a few small nits.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/npu_dma_to_half_dma_cpy_nd.mlir (hunk 3 above)
Comment (Yu-Zhewen) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/npu_dma_to_half_dma_cpy_nd.mlir]: thanks! done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td [hunk 4]
```diff
@@ -171,6 +171,7 @@ def AMDAIEControlCodeLowering :
   let constructor = "mlir::iree_compiler::AMDAIE::createAMDAIEControlCodeLoweringPass()";
 }
 
+
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td:174]:
```suggestion
```
Not sure why this new line?

**Review by jtuyls — APPROVED** [review]
LGTM

**Pull request merged**